storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java

```diff
@@ -0,0 +1,631 @@
+            signal(pid, 9, user);
+        } else {
+            LOG.warn("Trying to forceKill container {} but pidfile is not found", workerId);
+        }
+    }
+
+    // return null if not found.
+    private Long getContainerPid(String workerId) {
+        Long pid = workerToContainerPid.get(workerId);
+        if (pid == null) {
+            String containerPidFilePath = containerPidFile(workerId);
+            if (!new File(containerPidFilePath).exists()) {
+                LOG.warn("{} doesn't exist", containerPidFilePath);
+            } else {
+                try {
+                    pid = Long.parseLong(CgroupUtils.readFileByLine(containerPidFilePath).get(0));
+                    workerToContainerPid.put(workerId, pid);
+                } catch (IOException e) {
+                    LOG.warn("failed to read {}", containerPidFilePath);
+                }
+            }
+        }
+        return pid;
+    }
+
+    @Override
+    public void releaseResourcesForWorker(String workerId) {
+        super.releaseResourcesForWorker(workerId);
+        workerToContainerPid.remove(workerId);
+    }
+
+    /**
+     * The container terminates if any process inside the container dies.
+     * So we only need to check if the initial process is alive or not.
+     * @param user the user that the processes are running as
+     * @param workerId the id of the worker to kill
+     * @return true if all processes are dead; false otherwise
+     * @throws IOException on I/O exception
+     */
+    @Override
+    public boolean areAllProcessesDead(String user, String workerId) throws IOException {
+        boolean areAllDead = isContainerDead(workerId, user);
+        LOG.debug("WorkerId {}: Checking areAllProcessesDead: {}", workerId, areAllDead);
+        return areAllDead;
+    }
+
+    @Override
+    public void cleanup(String user, String workerId, int port) throws IOException {
+        LOG.debug("clean up worker {}", workerId);
+        try {
+            String containerId = getContainerId(workerId, port);
+            List<String> commands = Arrays.asList(CmdType.REAP_OCI_CONTAINER.toString(), containerId, String.valueOf(layersToKeep));
+            String logPrefix = "Worker Process " + workerId;
+            int result = ClientSupervisorUtils.processLauncherAndWait(conf, user, commands, null, logPrefix);
+            if (result != 0) {
+                LOG.warn("Failed cleaning up RuncWorker {}", workerId);
+            }
+        } catch (FileNotFoundException e) {
+            // This could happen if we had an IOException and failed launching the worker.
+            // We need to continue on in order for the worker directory to get cleaned up.
+            LOG.error("Failed to find container id for {} ({}), unable to reap container", workerId, e.getMessage());
+        }
+        //remove from the watched list
+        LOG.debug("Removing {} from the watched workers list", workerId);
+        workerToUser.remove(workerId);
```

Comment:
call releaseResourcesForWorker() instead?  

It looks like releaseResourcesForWorker() is called before cleanup() already.  

Do we need both?  

I didn't see a javadoc for cleanup().  I'm fuzzy on the difference.  I am all for combining into one if we can.

Reply:
makes sense. keeping `cleanup` and removing `releaseResourcesForWorker`

Reply:
Thanks for addressing this.